{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "58e751af",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "68e1d0b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "refseq_to_chm_number = {}\n",
    "\n",
    "with open(\"../../chm_to_acc_numbers.tsv\") as fh:\n",
    "    for line in fh:\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        line = line.strip().split(\"\\t\")\n",
    "        refseq_to_chm_number[line[6]] = line[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c7d3856",
   "metadata": {},
   "outputs": [],
   "source": [
    "#transcript\n",
    "\n",
    "## dict with search results\n",
    "\n",
    "# {read_n:{chr_n: {trans_n: str, chr_n: str, coordinates : [],total_length : int}}}\n",
    "\n",
    "reads_summary = defaultdict()\n",
    "\n",
    "## m8 == outftmt6 \n",
    "## query_seq_id => target_seq_id => chm_number => identity => length => mismatch => gaps => \n",
    "## ==> start_query => stop_query => start_target => stop_target => e_value => bit-score\n",
    "with open(\"../../human_transcript_pacbio_reads/result_long_reads_human_t2t_match.m8\") as fh:\n",
    "    for line in fh:\n",
    "        line = line.strip().split(\"\\t\")\n",
    "        read = line[0]\n",
    "        identity = float(line[2])\n",
    "        chr_n = refseq_to_chm_number[line[1]]\n",
    "        transcript = line[0].split('_')[4]\n",
    "        start = int(line[-4])\n",
    "        stop = int(line[-3])\n",
    "        start_read = int(line[-6])\n",
    "        stop_read = int(line[-5])\n",
    "        if start > stop:\n",
    "#             print('Stop < start, changing...')\n",
    "#             print(start, stop)\n",
    "            start, stop = stop, start\n",
    "#             print('Change to: ', start, stop)\n",
    "        if start_read > stop_read:\n",
    "            start_read, stop_read = stop_read, start_read\n",
    "        if ((stop - start) < 200) or (identity < 0.8):\n",
    "            continue\n",
    "        if not read in reads_summary.keys():\n",
    "            reads_summary[read] = {}\n",
    "        if not chr_n in reads_summary[read].keys():\n",
    "#             print('Adding initial coords to chr', chr_n, start, stop)\n",
    "            reads_summary[read][chr_n] = {\n",
    "                'transcript_n': transcript,\n",
    "                'chr_n': chr_n,\n",
    "                'coordinates_genome': [f\"{start}:{stop}\"],\n",
    "                'coordinates_read': [f\"{start_read}:{stop_read}\"],\n",
    "                'total_length_genome': stop - start,\n",
    "                'total_length_read' : stop_read - start_read\n",
    "            }\n",
    "        else:\n",
    "#             nested_longer = None\n",
    "#             nested_longer_index = None\n",
    "            not_nested = None\n",
    "            #difference = None\n",
    "            for i, coord in enumerate(reads_summary[read][chr_n]['coordinates_genome']):\n",
    "                start_in = int(coord.split(':')[0])\n",
    "                stop_in = int(coord.split(':')[1])\n",
    "#                 print(f'Coords check: {start_in}:{stop_in} to {start}:{stop}')\n",
    "#                 if start < start_in and stop > stop_in:\n",
    "#                     difference = (stop - start) - (stop_in - start_in)\n",
    "#                     nested_longer = True\n",
    "#                     nested_longer_index = i\n",
    "#                 elif start < start_in and start_in < stop < stop_in:\n",
    "#                     if (stop - start) > (stop_in - start_in):\n",
    "#                         difference = (stop - start) - (stop_in - start_in)\n",
    "#                         nested_longer = True\n",
    "#                         nested_longer_index = i\n",
    "#                 elif start_in < start < stop and stop > stop_in:\n",
    "#                     if (stop - start) > (stop_in - start_in):\n",
    "#                         difference = (stop - start) - (stop_in - start_in)\n",
    "#                         nested_longer = True\n",
    "#                         nested_longer_index = i\n",
    "                if (start < start_in) and (stop < start_in):\n",
    "#                     print('Not Nested 1! Continue check...')\n",
    "                    not_nested = True\n",
    "                elif (start > stop_in) and (stop > stop_in):\n",
    "                    not_nested = True\n",
    "#                     print('Not nested 2! Continue check...')\n",
    "                else:\n",
    "#                     print('Nested! Breaking...')\n",
    "                    not_nested = False\n",
    "                    break\n",
    "#             if nested_longer:\n",
    "#                 reads_summary[read][chr_n]['coordinates'][nested_longer_index] = f'{start}:{stop}'\n",
    "#                 reads_summary[read][chr_n]['total_length'] += difference\n",
    "            if not_nested:\n",
    "                reads_summary[read][chr_n]['coordinates_genome'].append(f'{start}:{stop}')\n",
    "                reads_summary[read][chr_n]['total_length_genome'] += stop - start\n",
    "                if f'{start_read}:{stop_read}' in reads_summary[read][chr_n]['coordinates_read']:\n",
    "                    continue\n",
    "                else:\n",
    "                    not_nested_read = None\n",
    "                    for i, coord in enumerate(reads_summary[read][chr_n]['coordinates_read']):\n",
    "                        start_in = int(coord.split(':')[0])\n",
    "                        stop_in = int(coord.split(':')[1])\n",
    "                        if (start_read < start_in) and (stop_read < start_in):\n",
    "                            not_nested_read = True\n",
    "                        elif (start_read > stop_in) and (stop_read > stop_in):\n",
    "                            not_nested_read = True\n",
    "                        else:\n",
    "                            not_nested_read = False\n",
    "                            break\n",
    "                    if not_nested_read:\n",
    "                        reads_summary[read][chr_n]['coordinates_read'].append(f'{start_read}:{stop_read}')\n",
    "                        reads_summary[read][chr_n]['total_length_read'] += stop_read - start_read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72621403",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('../data/reads_match_coords.json', 'w', encoding='utf-8') as f:\n",
    "    json.dump(reads_summary, f, ensure_ascii=False, indent=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fdfdd593",
   "metadata": {},
   "outputs": [],
   "source": [
    "reads_summary_without_coords = {}\n",
    "\n",
    "for read, chr_info in reads_summary.items():\n",
    "    print(read)\n",
    "    for chr_n, info in chr_info.items():\n",
    "        read_key = f\"{read}_chr_{chr_n}\"\n",
    "        print(read_key)\n",
    "        reads_summary_without_coords[read_key] = [chr_n, info['transcript_n'], str(info['total_length_genome']), str(info['total_length_read'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea631572",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/reads_match_summary.tsv', 'w') as fw:\n",
    "    fw.write('#read\\tchromosome\\ttranscript\\ttotal_length_genome\\ttotal_length_read\\n')\n",
    "    for read, summary in reads_summary_without_coords.items():\n",
    "        line_to_write = \"\\t\".join(summary)\n",
    "        fw.write(f\"{read}\\t{line_to_write}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "829e063e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
